File: android/nativepreview/src/main/java/com/travelplaner/nativepreview/domain/ExpenseSettlement.kt
package com.travelplaner.nativepreview.domain

data class SettlementParticipant(val id: String, val name: String)
data class SettlementExpense(
    val id: String,
    val amount: Int,
    val currency: String,
    val amountKrw: Int,
    val payerId: String,
    val participantIds: List<String>,
)
data class PersonSettlement(val id: String, val name: String, val paidKrw: Int, val shareKrw: Int, val balanceKrw: Int)
data class SettlementTransfer(val fromId: String, val toId: String, val amountKrw: Int)
data class SettlementSummary(val people: List<PersonSettlement>, val transfers: List<SettlementTransfer>, val totalKrw: Int, val expenseCount: Int)
data class CashWallet(val id: String, val name: String, val currency: String, val initial: Int = 0, val additional: Int = 0, val actualRemaining: Int? = null)
data class CashWalletRemovalState(val wallets: List<CashWallet>, val activeWalletId: String?, val cashLedgerCurrency: String)

object ExpenseSettlement {
    fun normalizeParticipants(source: List<SettlementParticipant>): List<SettlementParticipant> {
        val self = source.firstOrNull { it.id.trim() == "self" && it.name.trim().isNotEmpty() }?.let { SettlementParticipant("self", it.name.trim()) }
            ?: SettlementParticipant("self", "나")
        val result = mutableListOf(self)
        val seen = mutableSetOf("self")
        source.forEach { person ->
            val id = person.id.trim(); val name = person.name.trim()
            if (id.isNotEmpty() && name.isNotEmpty() && seen.add(id)) result += SettlementParticipant(id, name)
        }
        return result
    }

    fun normalizeExpense(expense: SettlementExpense, participants: List<SettlementParticipant>): SettlementExpense {
        val people = normalizeParticipants(participants)
        val known = people.map { it.id }.toSet()
        val payer = expense.payerId.trim().takeIf { it in known } ?: "self"
        val ids = expense.participantIds.filter { it in known }.distinct().toMutableList()
        if (ids.isEmpty()) ids += payer
        if (payer !in ids) ids += payer
        return expense.copy(amountKrw = maxOf(0, expense.amountKrw), payerId = payer, participantIds = ids)
    }

    fun calculate(expenses: List<SettlementExpense>, participants: List<SettlementParticipant>): SettlementSummary {
        val people = normalizeParticipants(participants)
        val paid = people.associate { it.id to 0 }.toMutableMap()
        val shares = people.associate { it.id to 0 }.toMutableMap()
        var total = 0; var count = 0
        expenses.forEach { raw ->
            val expense = normalizeExpense(raw, people)
            if (expense.amountKrw <= 0) return@forEach
            total += expense.amountKrw; count += 1
            paid[expense.payerId] = paid.getValue(expense.payerId) + expense.amountKrw
            val base = expense.amountKrw / expense.participantIds.size
            val remainder = expense.amountKrw - base * expense.participantIds.size
            expense.participantIds.forEachIndexed { index, id -> shares[id] = shares.getValue(id) + base + if (index < remainder) 1 else 0 }
        }
        val summaries = people.map { PersonSettlement(it.id, it.name, paid.getValue(it.id), shares.getValue(it.id), paid.getValue(it.id) - shares.getValue(it.id)) }
        val debtors = summaries.filter { it.balanceKrw < 0 }.map { it.id to -it.balanceKrw }.toMutableList()
        val creditors = summaries.filter { it.balanceKrw > 0 }.map { it.id to it.balanceKrw }.toMutableList()
        val transfers = mutableListOf<SettlementTransfer>()
        var debtorIndex = 0; var creditorIndex = 0
        while (debtorIndex < debtors.size && creditorIndex < creditors.size) {
            val amount = minOf(debtors[debtorIndex].second, creditors[creditorIndex].second)
            if (amount > 0) transfers += SettlementTransfer(debtors[debtorIndex].first, creditors[creditorIndex].first, amount)
            debtors[debtorIndex] = debtors[debtorIndex].first to debtors[debtorIndex].second - amount
            creditors[creditorIndex] = creditors[creditorIndex].first to creditors[creditorIndex].second - amount
            if (debtors[debtorIndex].second == 0) debtorIndex += 1
            if (creditors[creditorIndex].second == 0) creditorIndex += 1
        }
        return SettlementSummary(summaries, transfers, total, count)
    }

    fun createWallet(id: String, currency: String, name: String? = null): CashWallet {
        val safe = currency.trim().ifEmpty { "KRW" }
        return CashWallet(if (id.isBlank()) "$safe-cash-wallet" else id, name ?: "$safe 현금 지갑", safe)
    }

    fun removeWallet(wallets: List<CashWallet>, walletId: String, activeWalletId: String?, selectedCurrency: String, fallbackCurrency: String = "KRW"): CashWalletRemovalState {
        val remaining = wallets.filterNot { it.id == walletId }
        val active = remaining.firstOrNull { it.id == activeWalletId }
            ?: remaining.firstOrNull { it.currency == selectedCurrency }
            ?: remaining.firstOrNull()
        return CashWalletRemovalState(remaining, active?.id, active?.currency ?: selectedCurrency.ifBlank { fallbackCurrency })
    }
}

File: ios/NativeCore/Sources/NativeCore/ExpenseSettlement.swift
import Foundation

public struct SettlementParticipant: Codable, Equatable, Identifiable {
    public let id: String
    public let name: String

    public init(id: String, name: String) {
        self.id = id
        self.name = name
    }
}

public struct SettlementExpense: Codable, Equatable, Identifiable {
    public let id: String
    public let amount: Double
    public let currency: String
    public let amountKRW: Int
    public let payerID: String
    public let participantIDs: [String]

    public init(id: String, amount: Double, currency: String, amountKRW: Int, payerID: String, participantIDs: [String]) {
        self.id = id
        self.amount = amount
        self.currency = currency
        self.amountKRW = amountKRW
        self.payerID = payerID
        self.participantIDs = participantIDs
    }
}

public struct PersonSettlement: Codable, Equatable, Identifiable {
    public let id: String
    public let name: String
    public let paidKRW: Int
    public let shareKRW: Int
    public let balanceKRW: Int

    public init(id: String, name: String, paidKRW: Int, shareKRW: Int, balanceKRW: Int) {
        self.id = id
        self.name = name
        self.paidKRW = paidKRW
        self.shareKRW = shareKRW
        self.balanceKRW = balanceKRW
    }
}

public struct SettlementTransfer: Codable, Equatable {
    public let fromID: String
    public let toID: String
    public let amountKRW: Int

    public init(fromID: String, toID: String, amountKRW: Int) {
        self.fromID = fromID
        self.toID = toID
        self.amountKRW = amountKRW
    }
}

public struct SettlementSummary: Codable, Equatable {
    public let people: [PersonSettlement]
    public let transfers: [SettlementTransfer]
    public let totalKRW: Int
    public let expenseCount: Int

    public init(people: [PersonSettlement], transfers: [SettlementTransfer], totalKRW: Int, expenseCount: Int) {
        self.people = people
        self.transfers = transfers
        self.totalKRW = totalKRW
        self.expenseCount = expenseCount
    }
}

public struct CashWallet: Codable, Equatable, Identifiable {
    public let id: String
    public let name: String
    public let currency: String
    public var initial: Int
    public var additional: Int
    public var actualRemaining: Int?

    public init(id: String, name: String, currency: String, initial: Int = 0, additional: Int = 0, actualRemaining: Int? = nil) {
        self.id = id
        self.name = name
        self.currency = currency
        self.initial = initial
        self.additional = additional
        self.actualRemaining = actualRemaining
    }
}

public struct CashWalletRemovalState: Equatable {
    public let wallets: [CashWallet]
    public let activeWalletID: String?
    public let cashLedgerCurrency: String

    public init(wallets: [CashWallet], activeWalletID: String?, cashLedgerCurrency: String) {
        self.wallets = wallets
        self.activeWalletID = activeWalletID
        self.cashLedgerCurrency = cashLedgerCurrency
    }
}

public enum ExpenseSettlement {
    public static func normalizeParticipants(_ source: [SettlementParticipant]) -> [SettlementParticipant] {
        var result = [SettlementParticipant(id: "self", name: "나")]
        var seen = Set(["self"])
        if let selfEntry = source.first(where: { $0.id.trimmingCharacters(in: .whitespacesAndNewlines) == "self" }), !selfEntry.name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            result[0] = SettlementParticipant(id: "self", name: selfEntry.name.trimmingCharacters(in: .whitespacesAndNewlines))
        }
        for person in source {
            let id = person.id.trimmingCharacters(in: .whitespacesAndNewlines)
            let name = person.name.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !id.isEmpty, !name.isEmpty, !seen.contains(id) else { continue }
            seen.insert(id)
            result.append(SettlementParticipant(id: id, name: name))
        }
        return result
    }

    public static func normalizeExpense(_ expense: SettlementExpense, participants: [SettlementParticipant]) -> SettlementExpense {
        let people = normalizeParticipants(participants)
        let known = Set(people.map(\.id))
        let payer = known.contains(expense.payerID.trimmingCharacters(in: .whitespacesAndNewlines)) ? expense.payerID : "self"
        var ids: [String] = []
        for id in expense.participantIDs where known.contains(id) && !ids.contains(id) { ids.append(id) }
        if ids.isEmpty { ids = [payer] }
        if !ids.contains(payer) { ids.append(payer) }
        return SettlementExpense(id: expense.id, amount: expense.amount, currency: expense.currency, amountKRW: max(0, expense.amountKRW), payerID: payer, participantIDs: ids)
    }

    public static func calculate(expenses: [SettlementExpense], participants: [SettlementParticipant]) -> SettlementSummary {
        let people = normalizeParticipants(participants)
        var paid = Dictionary(uniqueKeysWithValues: people.map { ($0.id, 0) })
        var shares = Dictionary(uniqueKeysWithValues: people.map { ($0.id, 0) })
        var total = 0
        var count = 0

        for rawExpense in expenses {
            let expense = normalizeExpense(rawExpense, participants: people)
            guard expense.amountKRW > 0 else { continue }
            total += expense.amountKRW
            count += 1
            paid[expense.payerID, default: 0] += expense.amountKRW
            let base = expense.amountKRW / expense.participantIDs.count
            let remainder = expense.amountKRW - base * expense.participantIDs.count
            for (index, id) in expense.participantIDs.enumerated() {
                shares[id, default: 0] += base + (index < remainder ? 1 : 0)
            }
        }

        let summaries = people.map { person in
            PersonSettlement(id: person.id, name: person.name, paidKRW: paid[person.id] ?? 0, shareKRW: shares[person.id] ?? 0, balanceKRW: (paid[person.id] ?? 0) - (shares[person.id] ?? 0))
        }
        var debtors = summaries.filter { $0.balanceKRW < 0 }.map { ($0.id, amount: -$0.balanceKRW) }
        var creditors = summaries.filter { $0.balanceKRW > 0 }.map { ($0.id, amount: $0.balanceKRW) }
        var transfers: [SettlementTransfer] = []
        var debtorIndex = 0
        var creditorIndex = 0
        while debtorIndex < debtors.count && creditorIndex < creditors.count {
            let amount = min(debtors[debtorIndex].amount, creditors[creditorIndex].amount)
            if amount > 0 { transfers.append(SettlementTransfer(fromID: debtors[debtorIndex].0, toID: creditors[creditorIndex].0, amountKRW: amount)) }
            debtors[debtorIndex].amount -= amount
            creditors[creditorIndex].amount -= amount
            if debtors[debtorIndex].amount == 0 { debtorIndex += 1 }
            if creditors[creditorIndex].amount == 0 { creditorIndex += 1 }
        }
        return SettlementSummary(people: summaries, transfers: transfers, totalKRW: total, expenseCount: count)
    }

    public static func createWallet(id: String, currency: String, name: String? = nil) -> CashWallet {
        let safeCurrency = currency.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? "KRW" : currency
        return CashWallet(id: id.isEmpty ? "\(safeCurrency)-cash-wallet" : id, name: name ?? "\(safeCurrency) 현금 지갑", currency: safeCurrency)
    }

    public static func removeWallet(wallets: [CashWallet], walletID: String, activeWalletID: String?, selectedCurrency: String, fallbackCurrency: String = "KRW") -> CashWalletRemovalState {
        let remaining = wallets.filter { $0.id != walletID }
        let active = remaining.first(where: { $0.id == activeWalletID }) ?? remaining.first(where: { $0.currency == selectedCurrency }) ?? remaining.first
        return CashWalletRemovalState(wallets: remaining, activeWalletID: active?.id, cashLedgerCurrency: active?.currency ?? (selectedCurrency.isEmpty ? fallbackCurrency : selectedCurrency))
    }
}
